internal/batches/service/workspaces.go

```diff
@@ -0,0 +1,106 @@
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gobwas/glob"
+	batcheslib "github.com/sourcegraph/sourcegraph/lib/batches"
+	"github.com/sourcegraph/src-cli/internal/batches"
+	"github.com/sourcegraph/src-cli/internal/batches/graphql"
+)
+
+type RepoWorkspaces struct {
+	RepoID             string
+	Paths              []string
+	OnlyFetchWorkspace bool
+}
+
+// findWorkspaces matches the given repos to the workspace configs and
+// searches, via the Sourcegraph instance, the locations of the workspaces in
+// each repository.
+// The repositories that were matched by a workspace config are returned in
+// workspaces. root contains the repositories that didn't match a config.
+// If the user didn't specify any workspaces, the repositories are returned as
+// root repositories.
+func findWorkspaces(
+	ctx context.Context,
+	spec *batcheslib.BatchSpec,
+	finder directoryFinder,
+	repos []*graphql.Repository,
+) ([]RepoWorkspaces, error) {
+	// Pre-compile all globs.
```

Comment:
This function and type can relatively easy be pulled over to the backend now. Next step :)